benchmark/websocket/simple.js

```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+const common = require('../common.js');
+const http = require('http');
+const WebSocketServer = require('ws');
```

Comment:
I think this should not depend on `ws`, otherwise we have to add it as a dependency.

Reply:
Developing a WebSocket server without `ws` seems like a massive amount of work. Is there an alternative approach you would suggest?

Reply:
This is an echo example and we are benchmarking a client. Firstly we need to make sure that the server is faster than the client otherwise we might end up benchmarking the server. Secondly, instead of sending the initial frame from the client, we can prepare and send it from the server, and let the client send it back. By doing this we can simplify the server a lot. we don't need to parse frames on the server. After sending the initial frame we only count the number of received bytes on the server. When that number matches the expected length of the frame received from the client, we send the frame again and so on.

If it is not clear I will write a code example tomorrow. I need to sleep now.

Reply:
Thank you, I think I understand what you mean. The server sends a short initial frame each time, and the client replies with frames of 64/16k/128k sizes. The server then checks if the totalReceivedBytes has reached the expectedBytes, and upon fulfillment, it proceeds to send the next initial frame.

Reply:
Yes, here is an example.

```js
const crypto = require('crypto');
const http = require('http');

const GUID = '258EAFA5-E914-47DA-95CA-C5AB0DC85B11';

const server = http.createServer();

const frame = createFrame(Buffer.alloc(64).fill('.'), 1);
let roundtrips = 0;

server.on('upgrade', function (req, socket) {
  const key = crypto
    .createHash('sha1')
    .update(req.headers['sec-websocket-key'] + GUID)
    .digest('base64');

  let bytesReceived = 0;

  socket.on('data', function onData(chunk) {
    bytesReceived += chunk.length;

    if (bytesReceived === frame.length + 4) {
      // Message completely received.
      bytesReceived = 0;

      if (++roundtrips === 5000) {
        console.log('done');
        socket.removeListener('data', onData);
        socket.resume();
        socket.end();

        server.close();
      } else {
        socket.write(frame);
      }
    }
  });

  socket.write(
    [
      'HTTP/1.1 101 Switching Protocols',
      'Upgrade: websocket',
      'Connection: Upgrade',
      `Sec-WebSocket-Accept: ${key}`,
      '\r\n'
    ].join('\r\n')
  );

  socket.write(frame);
});

server.listen(8080, function () {
  const ws = new WebSocket('ws://localhost:8080');

  ws.addEventListener('message', function (event) {
    ws.send(event.data);
  });
});

function createFrame(data, opcode) {
  let offset = 2;
  let payloadLength = data.length;

  if (payloadLength >= 65536) {
    offset += 8;
    payloadLength = 127;
  } else if (payloadLength > 125) {
    offset += 2;
    payloadLength = 126;
  }

  const target = Buffer.alloc(offset);

  target[0] = opcode | 0x80;
  target[1] = payloadLength;

  if (payloadLength === 126) {
    target.writeUInt16BE(data.length, 2);
  } else if (payloadLength === 127) {
    target[2] = target[3] = 0;
    target.writeUIntBE(data.length, 4, 6);
  }

  return Buffer.concat([target, data]);
}
```

Reply:
Thank you for your generous help, I updated the server.